dali/operators/audio/preemphasis_filter_op.h

```diff
@@ -31,47 +38,55 @@ const int kNumOutputs = 1;
 template<typename Backend>
 class PreemphasisFilter : public Operator<Backend> {
  public:
-  ~PreemphasisFilter() override = default;
+  explicit PreemphasisFilter(const OpSpec &spec)
+      : Operator<Backend>(spec), output_type_(spec.GetArgument<DALIDataType>(arg_names::kDtype)) {}
 
+  ~PreemphasisFilter() override = default;
   DISABLE_COPY_MOVE_ASSIGN(PreemphasisFilter);
 
- protected:
-  explicit PreemphasisFilter(const OpSpec &spec) :
-          Operator<Backend>(spec),
-          output_type_(spec.GetArgument<std::remove_const_t<decltype(this->output_type_)>>(
-                  arg_names::kDtype)) {}
+  bool CanInferOutputs() const override {
+    return true;
+  }
 
+  bool SetupImpl(std::vector<::dali::OutputDesc> &output_desc,
+                 const workspace_t<Backend> &ws) override {
+    const auto &input = ws.template InputRef<Backend>(0);
+    AcquireArguments(ws);
+    output_desc.resize(detail::kNumOutputs);
+    auto shape = input.shape();
+    output_desc[0].shape = shape;
+    output_desc[0].type = TypeTable::GetTypeInfo(output_type_);
+
+    if (std::is_same<CPUBackend, Backend>::value) {
+      assert(sample_queue_.empty());
+      for (int sample_id = 0; sample_id < shape.num_samples(); sample_id++) {
+        sample_queue_.push({volume(shape[sample_id]), sample_id});
+      }
+    }
 
-  bool CanInferOutputs() const override {
     return true;
   }
 
+  void RunImpl(workspace_t<Backend> &ws) override;
+
+ private:
   void AcquireArguments(const ArgumentWorkspace &ws) {
     this->GetPerSampleArgument(preemph_coeff_, detail::kCoeff, ws);
   }
 
   USE_OPERATOR_MEMBERS();
   std::vector<float> preemph_coeff_;
   const DALIDataType output_type_;
-};
-
-
-class PreemphasisFilterCpu : public PreemphasisFilter<CPUBackend> {
- public:
-  explicit PreemphasisFilterCpu(const OpSpec &spec) : PreemphasisFilter(spec) {}
 
-  ~PreemphasisFilterCpu() override = default;
+  // Used for the GPU variant
+  Tensor<Backend> scratchpad_;
 
-  DISABLE_COPY_MOVE_ASSIGN(PreemphasisFilterCpu);
-
- protected:
-  bool SetupImpl(std::vector<::dali::OutputDesc> &output_desc,
-                 const workspace_t<CPUBackend> &ws) override;
-
-  void RunImpl(workspace_t<CPUBackend> &ws) override;
+  // Used for the CPU variant
+  using VolumeSampleIdPair = std::pair<int64_t, int>;  // volume(out_shape), sample_idx
+  using SampleQueue = std::priority_queue<VolumeSampleIdPair, std::vector<VolumeSampleIdPair>>;
+  SampleQueue sample_queue_;
```

Comment:
This suggests that CPU and GPU operators should be distinct classes. Depending on how big time pressure is, you may rework it or not.

Reply:
Done